File: Python/DPC.py
import numpy as np
from custom_rk45 import rk45_step
import matplotlib.pyplot as plt

# Pendulum Dynamics Model
def DPC_dynamics(state,static):
    # unpack args
    cart_properties, F, g = static

    # unpack state
    x, dx, t1, dt1, t2, dt2 = state

    # unpack properties
    m, m1, m2, l1, l2 = cart_properties

    # Math Credit: TU Berlin "Equations of motion for an inverted double pendulum on a cart (in generalized coordinates)"

    # calculate "M(y) matrix"
    My_matrix = np.zeros([3,3])

    My_matrix[0,0] = m + m1 + m1
    My_matrix[0,1] = l1*(m1 + m2)*np.cos(t1)
    My_matrix[0,2] = m2*l2*np.cos(t2)
    My_matrix[1,0] = l1 * (m1 + m2) * np.cos(t1)
    My_matrix[1,1] = l1**2 * (m1 + m2)
    My_matrix[1,2] = l1 * l2 * m2 * np.cos(t1 - t2)
    My_matrix[2,0] = l2 * m2 * np.cos(t2)
    My_matrix[2,1] = l1 * l2 * m2 * np.cos(t1 - t2)
    My_matrix[2,2] = l2**2 * m2

    # calculate RHS components
    RHS1 = np.zeros([3,1])

    RHS1[0] = l1 * (m1 + m2) * dt1**2 * np.sin(t1) + m2 * l2 * dt2**2 * np.sin(dt2)
    RHS1[1] = -l1 * l2 * m2 * dt2**2 * np.sin(t1 - t2) + g * (m1 + m2) * l1 * np.sin(t1)
    RHS1[2] = l1 * l2 * m2 * dt1**2 * np.sin(t1 - t2) + g * l2 * m2 * np.sin(t2)

    RHS2 = np.zeros([3,1])

    RHS2[0] = F

    # calculate ddx, ddt1 and ddt2
    result = np.matmul(np.linalg.inv(My_matrix),RHS1) + RHS2

    # return dstate

    return np.array([dx,float(result[0]),dt1,float(result[1]),dt2,float(result[2])])

# Define Cart
cart_properties = np.array([1,1,1,1,1]) # [m,m1,m2,l1,l2]

# Define physical constants
g = 9.81 # gravity
F = 0

static = [cart_properties,F,g]

# Define simulation
IC = np.array([0,0,0.1,0,0,0]) # [x, dx, t1, dt1, t2, dt2] where t is theta (rad)
max_time = 10 # seconds
steps = max_time * 1000
tvec = np.linspace(0,max_time,steps)
dt = max_time/steps

time_hist = np.zeros([steps,6])

cur_state = IC

# Run Simulation
for i, time in enumerate(tvec):
    cur_state = rk45_step(DPC_dynamics,cur_state,dt,static)
    time_hist[i,:] = cur_state

# plot results
plt.plot(tvec,time_hist[:,0])

plt.figure()
plt.plot(tvec,time_hist[:,2])
plt.plot(tvec,time_hist[:,4])

plt.show()

File: MATLAB/DPC.m
% Define IC and cart properties
IC = [0,-1,deg2rad(20),0,deg2rad(-20),0]; % [x,dx,t1,dt1,t2,dt2]
cart_properties = [0.1  ,1,1,1,3]; % [m,m1,m2,l1,l2]
g = 9.81;
F = 0;

% define simulation options
op = odeset('abstol',1e-13,'reltol',1e-13); % propegation options
tmax = 10; % seconds
steps = tmax*100;
t_vec = linspace(0,tmax,steps);
[t,state] = ode45(@(t, state) DPC_dynamics_model(cart_properties,state,g,F), t_vec, IC, op);

% generate frames
obj = VideoWriter("cart_animation");
open(obj);
for i = 1:length(t)
    draw_cart(state(i,:),cart_properties,[-5*cart_properties(4),5*cart_properties(4)])
    writeVideo(obj,getframe)
    clf
end
close(obj)

function dstate = DPC_dynamics_model(cart_properties,state,g,F)
% CREDIT FOR MATH: TU Berlin

% unpact cart properties
m = cart_properties(1);
m1 = cart_properties(2);
m2 = cart_properties(3);
l1 = cart_properties(4);
l2 = cart_properties(5);

% unpact cart state
x = state(1);
dx = state(2);
t1 = state(3);
dt1 = state(4);
t2 = state(5);
dt2 = state(6);

% calculate M(y) matrix
My = [m + m1 + m1, l1*(m1 + m2)*cos(t1), m2*l2*cos(t2);...
      l1 * (m1 + m2) * cos(t1), l1^2 * (m1 + m2), l1 * l2 * m2 * cos(t1 - t2);...
      l2 * m2 * cos(t2), l1 * l2 * m2 * cos(t1 - t2), l2^2 * m2];

% calculate RHS
RHS1 = [l1 * (m1 + m2) * dt1^2 * sin(t1) + m2 * l2 * dt2^2 * sin(dt2);...
       -l1 * l2 * m2 * dt2^2 * sin(t1 - t2) + g * (m1 + m2) * l1 * sin(t1);...
        l1 * l2 * m2 * dt1^2 * sin(t1 - t2) + g * l2 * m2 * sin(t2)];
RHS2 = [F;0;0];

% calculate vector of derivatives
result = inv(My) * RHS1 + RHS2;
ddx = result(1);
ddt1 = result(2);
ddt2 = result(3);

dstate = [dx;ddx;dt1;ddt1;dt2;ddt2];

end

% animator function
function draw_cart(state,cart_properties,bounds)
L1 = cart_properties(4);
L2 = cart_properties(5);

% plot cart
p1 = [state(1) - L1/2,L1/2];
p2 = [state(1) + L1/2,L1/2];
p3 = [state(1) + L1/2,-L1/2];
p4 = [state(1) - L1/2,-L1/2];
cart_points = vertcat(p1,p2,p3,p4,p1);
plot(cart_points(:,1),cart_points(:,2));

hold on

% other stuff
yline(0)
xline(0)

% plot pendulum 1
p6 = [state(1) + L1 * sin(state(3)),L1 * cos(state(3))];
pend_points = vertcat([state(1),0],p6,[state(1),0]);
plot(pend_points(:,1),pend_points(:,2))

% plot pendulum 2
p7 = [p6(1) + L2 * sin(state(5)),p6(2) + L2 * cos(state(5))];
pend_points = vertcat(p6,p7,p6);
plot(pend_points(:,1),pend_points(:,2))

axis equal
xlim(bounds)
ylim(bounds)
end